primitives/io/src/tasks.rs

```diff
@@ -0,0 +1,114 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parallel tasks.
+
+#[cfg(feature = "std")]
+mod inner {
+    use std::sync::mpsc;
+    use sp_externalities::ExternalitiesExt as _;
+    use crate::TaskExecutorExt;
+
+    /// Task handle (wasm).
+    ///
+    /// This can be `join`-ed to get (blocking) the result of
+    /// the spawned task execution.
+    pub struct DataJoinHandle {
+        receiver: mpsc::Receiver<Vec<u8>>,
+    }
+
+    impl DataJoinHandle {
+        /// Join handle returned by `spawn` function
+        pub fn join(self) -> Vec<u8> {
+            self.receiver.recv().expect("Essntial task failure: forked runtime terminated before sending result.")
+        }
+    }
+
+    /// Spawn new runtime task (native).
+    pub fn spawn(entry_point: fn(Vec<u8>) -> Vec<u8>, data: Vec<u8>) -> Result<DataJoinHandle, &'static str> {
+        let scheduler = sp_externalities::with_externalities(|mut ext| ext.extension::<TaskExecutorExt>()
+            .expect("No task executor associated with the current context!")
+            .clone()
+        ).ok_or("Spawn called outside of externalities context!")?;
+
+        let (sender, receiver) = mpsc::channel();
+        scheduler.spawn("parallel-runtime-spawn", Box::pin(async move {
+            let result = entry_point(data);
+            let _ = sender.send(result);
+        }));
+
+        Ok(DataJoinHandle { receiver })
+    }
+}
+
+#[cfg(not(feature = "std"))]
+mod inner {
+
+    use sp_std::vec::Vec;
+
+    /// Spawn new runtime task (wasm).
+    pub fn spawn(entry_point: fn(Vec<u8>) -> Vec<u8>, payload: Vec<u8>) -> DataJoinHandle {
+
+        /// Dynamic dispatch of wasm blob.
+        ///
+        /// Arguments are expected to be scale encoded in vector at address `payload_ptr` with length of
+        /// `payload_len`.
+        ///
+        /// Arguments: function pointer (u32), payload (Vec<u8>).
```

Comment:
Can we change the terminology a bit? Right now we are having a payload which consists of two parts: entry and... a payload.

We could rename this outer payload to say "input_data" or "encoded params" or something along these lines